overlord/state/change_test.go

```diff
@@ -1119,3 +1119,305 @@ func (ts *taskRunnerSuite) TestCheckTaskDependencies(c *C) {
+
+	// task1 (wait) => task2 (do) means need a reboot
+	t1.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+
+	// task1 (done) => task2 (wait) means need a reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+}
+
+func (cs *changeSuite) TestIsWaitingCircularDependency(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("change", "...")
+
+	t1 := st.NewTask("task1", "...")
+	t2 := st.NewTask("task2", "...")
+	t3 := st.NewTask("task3", "...")
+	t4 := st.NewTask("wait-task", "...")
+
+	// Setup circular dependency between t1,t2 and t3, they should
+	// still act normally.
+	t2.WaitFor(t1)
+	t3.WaitFor(t2)
+	t1.WaitFor(t3)
+
+	chg.AddTask(t1)
+	chg.AddTask(t2)
+	chg.AddTask(t3)
+	chg.AddTask(t4)
+
+	// To trigger the cyclic dependency check, we must trigger the isWaiting logic
+	// and we do this by putting t4 into WaitStatus.
+	t4.SetToWait(state.DoneStatus)
+
+	// task1 (do) => task2 (do) => task3 (do) no reboot
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) => task2 (do) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoingStatus)
+	c.Check(chg.Status(), Equals, state.DoingStatus)
+
+	// task1 (wait) => task2 (do) => task3 (do) means need a reboot
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+
+	// task1 (done) => task2 (wait) => task3 (do) means need a reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+}
+
+func (cs *changeSuite) TestIsWaitingMultipleDependencies(c *C) {
+	st := state.New(nil)
+	st.Lock()
+	defer st.Unlock()
+
+	chg := st.NewChange("change", "...")
+
+	t1 := st.NewTask("task1", "...")
+	t2 := st.NewTask("task2", "...")
+	t3 := st.NewTask("task3", "...")
+	t4 := st.NewTask("wait-task", "...")
+	t3.WaitFor(t1)
+	t3.WaitFor(t2)
+
+	chg.AddTask(t1)
+	chg.AddTask(t2)
+	chg.AddTask(t3)
+	chg.AddTask(t4)
+
+	// Put t4 into wait-status to trigger the isWaiting logic each time
+	// for the change.
+	t4.SetToWait(state.DoneStatus)
+
+	// task1 (do) + task2 (do) => task3 (do) no reboot
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (done) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (do) => task3 (do) no reboot
+	t1.SetStatus(state.DoneStatus)
+	t2.SetStatus(state.DoStatus)
+	c.Check(chg.Status(), Equals, state.DoStatus)
+
+	// task1 (done) + task2 (wait) => task3 (do) means need a reboot4
+	t1.SetStatus(state.DoneStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
+
+	// task1 (wait) + task2 (wait) => task3 (do) means need a reboot
+	t1.SetToWait(state.DoneStatus)
+	t2.SetToWait(state.DoneStatus)
+	c.Check(chg.Status(), Equals, state.WaitStatus)
```

Comment:
I tried here:
```
	t1.SetStatus(state.ErrorStatus)
	t2.SetToWait(state.DoneStatus)
	fmt.Println(chg.Status())

	t1.SetToWait(state.DoneStatus)
	t2.SetStatus(state.ErrorStatus)
	fmt.Println(chg.Status())
```
which printed:
```
Wait
Do
```
while I am just switching the statuses between t1 and t2, which are both dependencies for t3. This is related to my comment about the loop, I think.

Reply:
I just want to note, the expected results from your example is 

``` 
Do
Do
```